File: shells/androidtv/app/src/main/kotlin/games/couchpad/tinytrack/BlobStore.kt
package games.couchpad.tinytrack

import android.content.Context
import android.util.Log
import org.json.JSONArray
import org.json.JSONObject
import java.io.File

/**
 * Bytes kept between runs, in a directory, under names the ENGINE chooses.
 *
 * FOUR PRIMITIVES AND NO POLICY. This lists, reads, writes and deletes; it does
 * not know what a blob contains, what it is called, when it stops being valid or
 * which one to throw away. Those are rules rather than platform facts, and they
 * live in `libttp-runtime/ttp/blobstore.h` where they are stated once for every
 * shell and pinned by the `abi` ctest — because the answers are not equally
 * visible when wrong. A bad eviction wastes disk. A bad INVALIDATION serves a
 * stale blob forever, across restarts, with nothing on screen to say so.
 *
 * It knows nothing about WHAT it is storing either — not that a sun bake exists,
 * nor a silhouette layer, nor how many kinds there are. That is the point of the
 * walk in `ttp_display.h`: the engine lists its stores, decides what to read,
 * keep and drop, and hands this a name. There used to be a `BakeCache` beside
 * this file holding that choreography — the window a bake key is defined over,
 * whether the engine already had the bake, whether the build had actually baked
 * — and every line of it was knowledge about the ENGINE living in a shell, with
 * a mirror this side had to invalidate whenever a destroyed surface took the
 * renderer away. It is also why the SECOND blob kind needed no Kotlin at all.
 *
 * **[generation] is this shell's one real contribution**, and it is the install
 * time rather than the versionName. A `-dirty` build keeps one version string
 * across many edits, which is exactly the development loop where a stale blob is
 * most likely and hardest to spot; the install time moves whether the version
 * string did or not. The engine folds it into the NAME, so a new binary cannot
 * name the old one's file at all.
 *
 * NOTHING HERE IS LOAD-BEARING. A miss, a short read, a corrupt blob, a full
 * disk and a failed delete all mean the same thing: compute it again. That is
 * what a cache is, and it is why every road out of this file is a null or a
 * no-op rather than an exception.
 */
class BlobStore(context: Context, private val store: String) {

    private companion object { const val TAG = "BlobStore" }

    private val dir: File? = try {
        File(context.filesDir, store).apply { mkdirs() }
    } catch (e: Exception) {
        Log.w(TAG, "no $store directory; nothing will be cached", e)
        null
    }

    /** What identifies the binary that produced (or would produce) these bytes. */
    val generation: String = try {
        context.packageManager.getPackageInfo(context.packageName, 0).lastUpdateTime.toString()
    } catch (e: Exception) {
        // No identity is not "any identity": an empty generation would still
        // name a blob, and every build would then share it. Answer something
        // this run alone can match, so the cache simply never hits.
        Log.w(TAG, "no install stamp; caching for this run only", e)
        "unknown-" + System.nanoTime()
    }

    /**
     * Everything held, as the walk's `entries` argument.
     *
     * `usedMs` is this shell's clock and is never compared against anything but
     * its siblings — an epoch, an uptime and a file mtime all sort the same way.
     */
    fun entriesJson(): String {
        val entries = JSONArray()
        dir?.listFiles()?.forEach {
            if (!it.isFile) return@forEach
            // A `.part` is a write that died between its temp file and its
            // rename, so it is junk rather than a blob: it can never be read
            // back (no plan ever names it) but it WOULD count towards the cap,
            // and evicting a real blob to make room for a corpse is the one way
            // this could cost something. Dropped on sight, like [read] drops a
            // blob it cannot parse.
            if (it.name.endsWith(".part")) { runCatching { it.delete() }; return@forEach }
            entries.put(JSONObject().put("name", it.name).put("usedMs", it.lastModified().toDouble()))
        }
        return entries.toString()
    }

    /** The blob's bytes, or null for a miss. Touches it, so eviction sees it as used. */
    fun read(name: String): ByteArray? {
        val f = File(dir ?: return null, name)
        if (!f.isFile) return null
        return try {
            val bytes = f.readBytes()
            f.setLastModified(System.currentTimeMillis())
            bytes
        } catch (e: Exception) {
            Log.w(TAG, "unreadable blob $name; dropping it", e)
            f.delete()
            null
        }
    }

    /**
     * Store bytes under [name].
     *
     * Through a temporary, because a half-written blob that a crash leaves
     * behind must not be read back as a whole one — the engine would refuse it
     * on its length checks, but only after paying to find out.
     */
    fun write(name: String, bytes: ByteArray) {
        val d = dir ?: return
        try {
            val tmp = File(d, "$name.part")
            tmp.writeBytes(bytes)
            if (!tmp.renameTo(File(d, name))) tmp.delete()
        } catch (e: Exception) {
            Log.w(TAG, "could not store blob $name", e)
        }
    }

    /** Throw one away. A delete that fails is not an error; the plan repeats it. */
    fun delete(name: String) {
        val d = dir ?: return
        runCatching { File(d, name).delete() }
    }
}

/**
 * One [BlobStore] per store the ENGINE says it has.
 *
 * The names are asked for rather than typed, which is the whole point: this
 * shell does not know that a "bake" or a "mask" exists, only that the engine
 * keeps some kinds of derived bytes and that each kind wants its own directory.
 * A third kind needs no Kotlin at all.
 */
class BlobStores(context: Context) {

    private val stores: Map<String, BlobStore> =
        TtpJson.strings(Ttp.ttp_display_blob_stores())
            .associateWith { BlobStore(context, it) }

    /** Perform something for each store, with the name the engine calls it. */
    inline fun forEach(action: (BlobStore, String) -> Unit) {
        for ((name, store) in all) action(store, name)
    }

    /** Exposed for [forEach]'s inlining; not otherwise interesting. */
    val all: Map<String, BlobStore> get() = stores
}

File: shells/tvos/TinyTrackParty/Assets/BlobStore.swift
import Foundation

/// Bytes kept between runs, in a directory, under names the ENGINE chooses.
///
/// FOUR PRIMITIVES AND NO POLICY. This lists, reads, writes and deletes; it does
/// not know what a blob contains, what it is called, when it stops being valid
/// or which one to throw away. Those are rules rather than platform facts, and
/// they live in `libttp-runtime/ttp/blobstore.h` where they are stated once for
/// every shell and pinned by the `abi` ctest — because the answers are not
/// equally visible when wrong. A bad eviction wastes disk. A bad INVALIDATION
/// serves a stale blob forever, across restarts, with nothing on screen to say
/// so.
///
/// It knows nothing about what it is storing either. `ttp_display.h`'s walk
/// decides what to read, what to keep and what to drop, and hands over a name;
/// `ttp_display_blob_stores` says which kinds exist. A third kind needs no Swift.
///
/// **`generation` is this shell's one real contribution.** The engine folds it
/// into the NAME, so a new binary cannot name the old one's file at all — which
/// is what makes a stale blob unreachable rather than merely rejected. Android
/// uses the APK's install time; the web uses its BUILD_STAMP hash; here it is
/// the bundle's build number plus the executable's own mtime, because a
/// developer rebuild keeps one version string across many edits and the mtime
/// moves whether the version did or not.
///
/// NOTHING HERE IS LOAD-BEARING. A miss, a short read, a corrupt blob, a full
/// disk and a failed delete all mean the same thing: compute it again.
final class BlobStore {

    private let dir: URL?
    let generation: String

    init(store: String, generation: String) {
        self.generation = generation
        // CACHES, not Application Support: this is derived data the app can
        // rebuild at any time, and the system is entitled to reclaim it under
        // pressure — which is exactly the contract a cache wants. On tvOS that
        // matters more than on a phone, where storage is scarce by design.
        let caches = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first
        guard let base = caches?.appendingPathComponent("ttp-blobs/\(store)", isDirectory: true) else {
            dir = nil
            return
        }
        do {
            try FileManager.default.createDirectory(at: base, withIntermediateDirectories: true)
            dir = base
        } catch {
            print("[ttp] no \(store) blob directory; nothing will be cached: \(error)")
            dir = nil
        }
    }

    /// Everything held, as the walk's `entries` argument.
    ///
    /// `usedMs` is this shell's clock and is never compared against anything but
    /// its siblings — an epoch, an uptime and a file mtime all sort the same way.
    func entriesJSON() -> String {
        guard let dir else { return "[]" }
        let names = (try? FileManager.default.contentsOfDirectory(
            at: dir, includingPropertiesForKeys: [.contentModificationDateKey])) ?? []
        var entries: [[String: Any]] = []
        for url in names {
            // A `.part` is a write that died between its temp file and its
            // rename, so it is junk rather than a blob: no plan can ever name
            // it, but it WOULD count towards the cap, and evicting a real blob
            // to make room for a corpse is the one way this could cost
            // something. Dropped on sight.
            if url.lastPathComponent.hasSuffix(".part") {
                try? FileManager.default.removeItem(at: url)
                continue
            }
            let stamp = (try? url.resourceValues(forKeys: [.contentModificationDateKey]))?
                .contentModificationDate?.timeIntervalSince1970 ?? 0
            entries.append(["name": url.lastPathComponent, "usedMs": stamp * 1000])
        }
        return TTP.json(entries)
    }

    /// The blob's bytes, or nil for a miss. Touches it, so eviction sees it used.
    func read(_ name: String) -> Data? {
        guard let dir else { return nil }
        let url = dir.appendingPathComponent(name)
        guard let data = try? Data(contentsOf: url) else { return nil }
        try? FileManager.default.setAttributes([.modificationDate: Date()], ofItemAtPath: url.path)
        return data
    }

    /// Store bytes under `name`, through a temporary — a half-written blob that
    /// a crash leaves behind must not be read back as a whole one. The engine
    /// would refuse it on its length checks, but only after paying to find out.
    func write(_ name: String, _ bytes: Data) {
        guard let dir else { return }
        let tmp = dir.appendingPathComponent("\(name).part")
        let dst = dir.appendingPathComponent(name)
        do {
            try bytes.write(to: tmp, options: .atomic)
            try? FileManager.default.removeItem(at: dst)
            try FileManager.default.moveItem(at: tmp, to: dst)
        } catch {
            try? FileManager.default.removeItem(at: tmp)
            print("[ttp] could not store blob \(name): \(error)")
        }
    }

    /// Throw one away. A delete that fails is not an error; the plan repeats it.
    func delete(_ name: String) {
        guard let dir else { return }
        try? FileManager.default.removeItem(at: dir.appendingPathComponent(name))
    }
}

/// One `BlobStore` per store the ENGINE says it has.
///
/// The names are asked for rather than typed, which is the point: this shell
/// does not know that a bake or a mask exists, only that the engine keeps some
/// kinds of derived bytes and that each kind wants its own directory.
final class BlobStores {

    private let stores: [(name: String, store: BlobStore)]

    init() {
        // What identifies this binary. Shared by every store, because it
        // describes the builder rather than the built.
        let info = Bundle.main.infoDictionary
        let version = (info?["CFBundleVersion"] as? String) ?? "0"
        let exe = Bundle.main.executableURL
        let mtime = (try? exe?.resourceValues(forKeys: [.contentModificationDateKey]))?
            .contentModificationDate?.timeIntervalSince1970 ?? 0
        let generation = "\(version)-\(Int(mtime))"
        stores = TTP.arr(ttp_display_blob_stores()).compactMap { $0 as? String }
            .map { ($0, BlobStore(store: $0, generation: generation)) }
    }

    /// Perform something for each store, with the name the engine calls it.
    func forEach(_ action: (BlobStore, String) -> Void) {
        for entry in stores { action(entry.store, entry.name) }
    }
}
